### ISSUE TITLE
rails 4.2

### ISSUE THREAD

**Issue by apneadiving** [oelmekki/activerecord_any_of#28]
seems we have an incompatibility there :)

**Comment by apneadiving** [comment]
to be clearer, I think binding rules have changed since the generated sql contains a `$1`:

    AND (NOT (\"users\".\"status\" = $1 AND (\"users\".\"email\" IS NOT NULL)))

**Comment by apneadiving** [comment]
asked the question to tenderlove about `where_values` potential bug https://twitter.com/apneadiving/status/553338758842818560

might be related

**Comment by oelmekki** [comment]
Hi Ben,

Thanks for reporting it !

Sorry, I haven't migrated to rails-4.2 on my main app because of an
other annoying issue we have (`rake db:schema:load` will now drop the
database and recreate it, which, even if our db user is given db
creation rights, prevent to recreate all postgres extensions since only
superuser can create those containing C code and make our CI
systematically fails).

I'll take a look at this this week-end on a dedicated app.

**Comment by apneadiving** [comment]
I have just been thrown away reporting as an issue on Rails: https://github.com/rails/rails/issues/18481

is it me or an "internal" should be private? ...

**Comment by matthewd** [comment]
@apneadiving, "I know this is explicitly not supported, but will you please provide me some support" doesn't seem to have many other likely outcomes.

No-one here is unclear on whether `where_values` is internal. If it were private, this gem would still need it, and would just use `send`. (And to be clear: there's nothing wrong with doing that. One just shouldn't expect help when it breaks.)

Back on topic: @oelmekki, I suspect this will be straight-forward, but if it proves tricky, feel free to ping me... I'll presumably need to deal with the same issue in the near future for rails/rails#16052, and can probably wrangle some unofficial support thereby.

**Comment by apneadiving** [comment]
@matthewd funny github links stuff, we dont have the same point of view about public methods, I read in another issue you consider it not official, thats why I added this iot prove I made some efforts. But no debate here: as maintainer you are right.

**Comment by nurey** [comment]
also encountering this issue. would love to see this updated with rails 4.2 support.

**Comment by oelmekki** [comment]
Hi there,

Sorry for taking so long, I'm beginning to look at it.

Could you provide a ruby statement that triggers your problem ?

I've tested with a rails-4.2 app using sqlite without any problem. 
Since there have been mentions of prepared statements, I suppose it's 
where the problem resides (especially since activerecord_any_of does 
special trickery around them).

**Comment by kreintjes** [comment]
@oelmekki  Hi,

I can confirm this problem. I am running Rails 4.2 with activerecord_any_of 1.2 and database PostgreSQL (using pg gem 0.18.1). All any_of queries using bind parameters result in a `PG::ProtocolViolation: ERROR:  bind message supplies 0 parameters, but prepared statement "" requires n` error. Some examples that do not work:

```
Comment.where.any_of(id: 3)
Comment.where.any_of({ id: 3 }, { id: 4 })
Comment.where.any_of(Comment.where(id: 3), Comment.where(id: 4))
```

However, queries like the following do work:

```
Comment.where.any_of('id = 3')
Comment.where.any_of('id = 3', 'id = 4')
```

**Comment by xtagon** [comment]
Hi, also encountering this issue in Rails 4.2.0 with activerecord_any_of 1.2 and PostgreSQL. Any updates on this issue?

**Comment by oelmekki** [comment]
Hi everyone,

Teaser : you gonna hate me

Sorry it took so long, I'm currently working on two companies and 
didn't encountered that bug myself, despite using postgresql and 
prepared statements.

I finally decided to dedicate some time to it and realized ... I did
not have this problem because it has been fixed for a long time on
master (in #20 ). Just forgot to release it as gem (#fail).

It's now published.
